pq-crypto/bike_r3/decode_avx2.c

```diff
@@ -0,0 +1,173 @@
+    in256             = BLENDV_I8(carry_in, carry_out, zero_mask2);
+
+    // Shift less than 64 (quadwords internal)
+    __m256i inner_carry = BLENDV_I8(carry_in, in256, zero_mask);
+    inner_carry         = PERM_I64(inner_carry, 0x39);
+    const __m256i out256 =
+      SRLI_I64(in256, count64) | SLLI_I64(inner_carry, (int)64 - count64);
+
+    // Store the rotated value
+    STORE(&out->qw[4 * i], out256);
+    carry_in = carry_out;
+  }
+}
+
+void rotate_right_avx2(OUT syndrome_t *out,
+                       IN const syndrome_t *in,
+                       IN const uint32_t    bitscount)
+{
+  // 1) Rotate in granularity of 256 bits blocks, using YMMs
+  rotate256_big(out, in, (bitscount / BITS_IN_YMM));
+  // 2) Rotate in smaller granularity (less than 256 bits), using YMMs
+  rotate256_small(out, out, (bitscount % BITS_IN_YMM));
+}
+
+// Duplicates the first R_BITS of the syndrome three times
+// |------------------------------------------|
+// |  Third copy | Second copy | first R_BITS |
+// |------------------------------------------|
+// This is required by the rotate functions.
+void dup_avx2(IN OUT syndrome_t *s)
+{
+  s->qw[R_QWORDS - 1] =
+    (s->qw[0] << LAST_R_QWORD_LEAD) | (s->qw[R_QWORDS - 1] & LAST_R_QWORD_MASK);
+
+  for(size_t i = 0; i < (2 * R_QWORDS) - 1; i++) {
+    s->qw[R_QWORDS + i] =
+      (s->qw[i] >> LAST_R_QWORD_TRAIL) | (s->qw[i + 1] << LAST_R_QWORD_LEAD);
+  }
+}
+
+// Use half-adder as described in [1].
+void bit_sliced_adder_avx2(OUT upc_t *upc,
+                           IN OUT syndrome_t *rotated_syndrome,
+                           IN const size_t    num_of_slices)
+{
+  // From cache-memory perspective this loop should be the outside loop
+  for(size_t j = 0; j < num_of_slices; j++) {
+    for(size_t i = 0; i < R_QWORDS; i++) {
+      const uint64_t carry = (upc->slice[j].u.qw[i] & rotated_syndrome->qw[i]);
+      upc->slice[j].u.qw[i] ^= rotated_syndrome->qw[i];
+      rotated_syndrome->qw[i] = carry;
+    }
+  }
+}
+
+void bit_slice_full_subtract_avx2(OUT upc_t *upc, IN uint8_t val)
+{
+  // Borrow
+  uint64_t br[R_QWORDS] = {0};
+
+  for(size_t j = 0; j < SLICES; j++) {
+
+    const uint64_t lsb_mask = 0 - (val & 0x1);
+    val >>= 1;
+
+    // Perform a - b with c as the input/output carry
+    // br = 0 0 0 0 1 1 1 1
+    // a  = 0 0 1 1 0 0 1 1
+    // b  = 0 1 0 1 0 1 0 1
+    // -------------------
+    // o  = 0 1 1 0 0 1 1 1
+    // c  = 0 1 0 0 1 1 0 1
+    //
+    // o  = a^b^c
+    //            _     __    _ _   _ _     _
+    // br = abc + abc + abc + abc = abc + ((a+b))c
+
+    for(size_t i = 0; i < R_QWORDS; i++) {
+      const uint64_t a      = upc->slice[j].u.qw[i];
+      const uint64_t b      = lsb_mask;
+      const uint64_t tmp    = ((~a) & b & (~br[i])) | ((((~a) | b) & br[i]));
+      upc->slice[j].u.qw[i] = a ^ b ^ br[i];
+      br[i]                 = tmp;
+    }
+  }
+}
+
+#endif
+
+typedef int dummy_typedef_to_avoid_empty_translation_unit_warning;
```

Comment:
The advantage is simpler logic in Makefile/CMake, I think we made them complicated enough with PQ algorithms so I just wanted to avoid further complications :)